src/Symfony/Component/HttpClient/NativeHttpClient.php

```diff
@@ -0,0 +1,409 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\HttpClient;
+
+use Symfony\Component\HttpClient\Exception\TransportException;
+use Symfony\Component\HttpClient\Response\NativeResponse;
+use Symfony\Component\HttpClient\Response\ResponseStream;
+use Symfony\Contracts\HttpClient\HttpClientInterface;
+use Symfony\Contracts\HttpClient\ResponseInterface;
+use Symfony\Contracts\HttpClient\ResponseStreamInterface;
+
+/**
+ * A portable implementation of the HttpClientInterface contracts based on PHP stream wrappers.
+ *
+ * PHP stream wrappers are able to fetch response bodies concurrently,
+ * but each request is opened synchronously.
+ *
+ * @author Nicolas Grekas <p@tchwork.com>
+ *
+ * @experimental in 4.3
+ */
+final class NativeHttpClient implements HttpClientInterface
+{
+    use HttpClientTrait;
+
+    private $defaultOptions = self::OPTIONS_DEFAULTS;
+    private $multi;
+
+    /**
+     * @param array $defaultOptions     Default requests' options
+     * @param int   $maxHostConnections The maximum number of connections to open
+     */
+    public function __construct(array $defaultOptions = [], int $maxHostConnections = 6)
+    {
+        if ($defaultOptions) {
+            [, $this->defaultOptions] = self::prepareRequest(null, null, $defaultOptions, self::OPTIONS_DEFAULTS);
+        }
+
+        // Use an internal stdClass object to share state between the client and its responses
+        $this->multi = (object) [
+            'openHandles' => [],
+            'handlesActivity' => [],
+            'pendingResponses' => [],
+            'maxHostConnections' => $maxHostConnections,
+            'responseCount' => 0,
+            'dnsCache' => [],
+            'handles' => [],
+            'sleep' => false,
+            'id' => random_int(PHP_INT_MIN, PHP_INT_MAX),
+        ];
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function request(string $method, string $url, array $options = []): ResponseInterface
```

Comment:
```php
    /**
     * @see HttpClientInterface::OPTIONS_DEFAULTS for $options
     *
     * {@inheritdoc}
     */
```

And I would do this on all the `request()` methods. We can't natively add auto-completion for the options. But, for devs like me, I immediately cmd+clicked into this method when I was using it. Buy adding this, I immediately can see the `@see` and can even Cmd+click it to go directly to that const.

Reply:
Maybe we could also add all known option keys as constants somewhere, like for instance Guzzle does.

https://github.com/guzzle/guzzle/blob/fa745406c2f48210693ba05f11280a1473037b2f/src/RequestOptions.php#L121-L125

Reply:
Good idea, added. About using constants, I proposed HttpOptions instead, which allows adding inline validation if we want to later on. I'm not 100% sure HttpOptions is the definitive answer on the topic - and neither am I that constants are preferable. Can be considered in a future PR.